define not working with typescript vue template

### Describe the bug

Using the vue-ts template, and defining a variable in the define section of the vite config, throws an error when building (works in dev)


### Reproduction

create a new vite vue-ts template project
```
$ pnpm create vite my-vue-ts-app -- --template vue-ts`
$ cd my-vue-ts-app
$ pnpm i
```

vite.config.ts
```js
export default defineConfig({
  define: {
    __BUILD_DATE: JSON.stringify(Date())
  },
  plugins: [vue()]
})
```

main.ts
```
import { createApp } from 'vue'
import App from './App.vue'

createApp(App).mount('#app')

console.log(__BUILD_DATE)
```

and then run build
```
➜  my-vue-ts-app pnpm run build

> my-vue-ts-app@0.0.0 build /Users/blah/Documents/test/vite/my-vue-ts-app
> vue-tsc --noEmit && vite build

src/main.ts:6:13 - error TS2304: Cannot find name '__BUILD_DATE'.

6 console.log(__BUILD_DATE)
              ~~~~~~~~~~~~


Found 1 error.

 ELIFECYCLE  Command failed with exit code 2.
```


### System Info

```shell
System:
    OS: macOS 11.6
    CPU: (16) x64 Intel(R) Core(TM) i9-9880H CPU @ 2.30GHz
    Memory: 336.57 MB / 16.00 GB
    Shell: 5.8 - /bin/zsh
  Binaries:
    Node: 17.2.0 - ~/.nvm/versions/node/v17.2.0/bin/node
    npm: 8.2.0 - ~/.nvm/versions/node/v17.2.0/bin/npm
  Browsers:
    Brave Browser: 95.1.31.88
    Chrome: 96.0.4664.93
    Edge: 96.0.1054.43
    Firefox: 94.0.2
    Safari: 15.0
  npmPackages:
    @vitejs/plugin-vue: ^1.10.2 => 1.10.2
    vite: ^2.7.0 => 2.7.1
```


### Used Package Manager

pnpm

### Logs

```shell
Same as above although I realise now that I use `pnpm run build` which actually runs `vue-tsc --noEmit && vite build` so it's actually vue-tsc that is failing not vite build...
```


### Validations

- [X] Follow our [Code of Conduct](https://github.com/vitejs/vite/blob/main/CODE_OF_CONDUCT.md)
- [X] Read the [Contributing Guidelines](https://github.com/vitejs/vite/blob/main/CONTRIBUTING.md).
- [X] Read the [docs](https://vitejs.dev/guide).
- [X] Check that there isn't [already an issue](https://github.com/vitejs/vite/issues) that reports the same bug to avoid creating a duplicate.
- [X] Make sure this is a Vite issue and not a framework-specific issue. For example, if it's a Vue SFC related bug, it should likely be reported to https://github.com/vuejs/vue-next instead.
- [X] Check that this is a concrete bug. For Q&A open a [GitHub Discussion](https://github.com/vitejs/vite/discussions) or join our [Discord Chat Server](https://chat.vitejs.dev/).
- [X] The provided reproduction is a [minimal reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) of the bug.

--- a/docs/config/index.md
+++ b/docs/config/index.md
@@ -131,9 +131,14 @@ export default defineConfig(async ({ command, mode }) => {
 
   For example, `process.env.FOO` and `__APP_VERSION__` are good fits. But `process` or `global` should not be put into this option. Variables can be shimmed or polyfilled instead.
 
+  ::: tip NOTE
+  For using with TypeScript, make sure to add the type declarations in the
+  `env.d.ts` or `vite-env.d.ts` file in order to pass type checking.
+  :::
+
 ### plugins
 
-- **Type:** ` (Plugin | Plugin[])[]`
+- **Type:** `(Plugin | Plugin[])[]`
 
   Array of plugins to use. Falsy plugins are ignored and arrays of plugins are flattened. See [Plugin API](/guide/api-plugin) for more details on Vite plugins.
 

--- a/docs/config/index.md
+++ b/docs/config/index.md
@@ -132,8 +132,13 @@ export default defineConfig(async ({ command, mode }) => {
   For example, `process.env.FOO` and `__APP_VERSION__` are good fits. But `process` or `global` should not be put into this option. Variables can be shimmed or polyfilled instead.
 
   ::: tip NOTE
-  For using with TypeScript, make sure to add the type declarations in the
-  `env.d.ts` or `vite-env.d.ts` file in order to pass type checking.
+  For TypeScript users, make sure to add the type declarations in the `env.d.ts` or `vite-env.d.ts` file to get type checks and Intellisense.
+
+  Example:
+  ```ts
+  // vite-env.d.ts
+  declare const __APP_VERSION__: string;
+  ```
   :::
 
 ### plugins

